cpp/modmesh/toggle/RadixTree.cpp

```diff
@@ -72,6 +72,169 @@ void CallProfiler::print_profiling_result(const RadixTreeNode<CallerProfile> & n
+        }
+
+        // Newline after the last element.
+        outstream << '\n';
+        outstream << R"(        },)" << '\n';
+    }
+}
+
+void CallProfilerSerializer::serialize_radix_tree_nodes(const RadixTreeNode<CallerProfile> * node, std::ostream & outstream)
+{
+    // Serialize all the RadixTreeNodes in RadixTree in a breadth-first manner.
+    outstream << R"(        "nodes": [)";
+
+    // Give each node a unique number
+    int unique_node_number = -1;
+
+    std::queue<const RadixTreeNode<CallerProfile> *> nodes_buffer;
+    CallProfilerSerializer::node_to_number_map_type node_to_unique_number;
+
+    nodes_buffer.push(node);
+    node_to_unique_number[node] = unique_node_number;
+    bool is_first_node = true;
+
+    // BFS algorithm
+    while (!nodes_buffer.empty())
+    {
+        const int nodes_buffer_size = nodes_buffer.size();
+        for (int i = 0; i < nodes_buffer_size; ++i)
+        {
+            const RadixTreeNode<CallerProfile> * current_node = nodes_buffer.front();
+            nodes_buffer.pop();
+
+            for (const auto & child : current_node->children())
+            {
+                ++unique_node_number;
+                nodes_buffer.push(child.get());
+                node_to_unique_number[child.get()] = unique_node_number;
+            }
+
+            CallProfilerSerializer::serialize_radix_tree_node(*current_node, is_first_node, node_to_unique_number, outstream);
+            is_first_node = false;
+
+            // Remove the node from the map
+            node_to_unique_number.erase(current_node);
+        }
+    }
+
+    // Newline after the last element.
+    outstream << '\n';
+    outstream << R"(        ])" << '\n';
+}
+
+void CallProfilerSerializer::serialize_radix_tree_node(const RadixTreeNode<CallerProfile> & node, bool is_first_node, CallProfilerSerializer::node_to_number_map_type & node_to_unique_number, std::ostream & outstream)
+{
+    // Serialize the RadixTreeNode to the json format.
+
+    // Avoid the trailing comma for the first node.
+    if (!is_first_node)
+    {
+        outstream << R"(,)" << '\n';
+        outstream << R"(            {)" << '\n';
+    }
+    else
+    {
+        outstream << R"({)" << '\n';
+    }
+
+    outstream << R"(                "unique_number": )" << node_to_unique_number[&node] << R"(,)" << '\n';
+    outstream << R"(                "key": )" << node.key() << R"(,)" << '\n';
+    outstream << R"(                "name": ")" << node.name() << R"(",)" << '\n';
+    CallProfilerSerializer::serialize_caller_profile(node.data(), outstream);
+    CallProfilerSerializer::serialize_radix_tree_node_children(node.children(), node_to_unique_number, outstream);
+
+    outstream << R"(            })";
+}
+
+void CallProfilerSerializer::serialize_radix_tree_node_children(const CallProfilerSerializer::child_list_type & children, CallProfilerSerializer::node_to_number_map_type & node_to_unique_number, std::ostream & outstream)
```

Comment:
This line is 216 char long.  It is too long.  But let's leave it here for now.  It is hard to set a threshold for how long a line should be, and it is subjective.  I do not want to set a number yet.

For a maintainer to quickly parse the argument by naked eyes, (i.e., readability's sake), the code style for this may be like (not clang-formatted):

```cpp
void CallProfilerSerializer::serialize_radix_tree_node(
    const RadixTreeNode<CallerProfile> & node, bool is_first_node,
    CallProfilerSerializer::node_to_number_map_type & node_to_unique_number,
    std::ostream & outstream)
```

Reply:
> This line is 216 char long. It is too long. But let's leave it here for now. It is hard to set a threshold for how long a line should be, and it is subjective. I do not want to set a number yet.
> 
> For a maintainer to quickly parse the argument by naked eyes, (i.e., readability's sake), the code style for this may be like (not clang-formatted):
> 
> ```c++
> void CallProfilerSerializer::serialize_radix_tree_node(
>     const RadixTreeNode<CallerProfile> & node, bool is_first_node,
>     CallProfilerSerializer::node_to_number_map_type & node_to_unique_number,
>     std::ostream & outstream)
> ```

I think the bad naming style results in the excessive long line, 
but I have no idea how to name a type(by using or typedef)/varialble/function for a complex hierarchical structure.
Do you have any suggestions?

Reply:
Long names are inevitable.  It is not practical to find a general way to name things perfectly when the system is complex (like to have 100 classes and 3,000 functions).

The only practical approach I know is to constantly review and refactor.